clang/test/CodeGenCXX/pod-member-memcpys.cpp

```diff
@@ -185,6 +203,18 @@ CALL_AO(PackedMembers)
 // CHECK: call void @llvm.memcpy.p0.p0.i64({{.*}} align 1 {{.*}} align 1 {{.*}}i64 16, i1 {{.*}})
 // CHECK: ret ptr
 
+// WithEmptyField copy-assignment:
+// CHECK-LABEL: define linkonce_odr nonnull align {{[0-9]+}} dereferenceable({{[0-9]+}}) ptr @_ZN14WithEmptyFieldaSERKS_
+// CHECK: call void @llvm.memcpy.p0.p0.i64({{.*}} align 4 {{.*}} align 4 {{.*}}i64 4, i1 {{.*}})
```

Comment:
@efriedma-quic Before this patch, this would've been a `memcpy` of size `5`. But the structure layout for `WithEmptyField` changed from:
```
%struct.WithEmptyField = type { i32, %struct.Empty, %struct.NonPOD, i32 }
%struct.Empty = type { i8 }                                   
%struct.NonPOD = type { i16 }                                 
```
to
```
struct.Foo = type { i32, %struct.NonPOD, i32 }
struct.NonPOD = type { i16 }                  
```
As you said, if we kept the `isZeroSize` check when generating this `memcpy` we would've written too much.

Reply:
Sorry, jumping in here without a lot of context, but the change in `%struct.WithEmptyField` seems wrong, doesn't it? Like that is a change in layout/ABI break/bad thing? I guess I'm missing something about why that change in layout would be acceptable? (maybe padding - if NonPOD gets aligned, such that the structure isn't changing layout after all?)

Reply:
+1, at least looking at the LLVM struct types that you pasted here, the offset of NonPOD in WithEmptyField has changed from 6 to 4, and that seems incorrect.

Reply:
I assume that isn't actually what's happening, and there's somehow a typo in the comment.

Reply:
Uhm yea that would be an ABI break, and I'm not sure how I got to this IR, apologies for the confusion. What actually happens is that we get:
```
%struct.WithEmptyField = type { i32, [4 x i8], i32 }
```
which seems fine. The `memcpy` does change from `5` to 4`